File: medicalcheckrest/WMedicalCheck/WMedicalCheck/medicalcheck/src/main/java/ch/hearc/medicalcheck/service/MeasureService.java
package ch.hearc.medicalcheck.service;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import javax.persistence.Tuple;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import ch.hearc.medicalcheck.model.Measure;
import ch.hearc.medicalcheck.repository.MeasureRepository;

/*
* Project   : Medical Check Rest
* Authors   : William Bikuta, Milán Cerviño, Ilyas Boillat, David Oktay
* Date      : 28.01.2022
* Class     : INF3dlm-a
* */

/**
 * bridge between repository and controller
 * each methods is explained in the associate controller
 * check them for more information
 */

@Service
public class MeasureService {
	@Autowired
	private MeasureRepository measureRepository;

	public List<Measure> getAll() {
		return measureRepository.findAll();
	}

	public Measure get(Integer id) {
		return measureRepository.findById(id).get();
	}

	public List<Measure> getAllByIdUser(Integer iduser) {
		return measureRepository.getAllMeasureByIdUser(iduser);
	}
	
	public Measure getLastByIdUser(Integer iduser) {
		return measureRepository.getLastMeasureByIdUser(iduser);
	}
	
	public Map<Integer, Double> getMapAverage(Integer iduser, String date) {
		
		List<Tuple> list = measureRepository.getMapAverageMeasure(iduser, date);
		return list.stream().collect(Collectors.toMap(t -> (Integer)t.get(0), t -> (Double)t.get(1)));
	}

	public Measure create(Measure measure) {
		return measureRepository.save(measure);
	}

	public Measure update(Integer idMeasure, Measure measure) {
		Optional<Measure> measureData = measureRepository.findById(measure.getId());

		if (measureData.isPresent()) {
			measure.setId(idMeasure);
			return measureRepository.save(measure);
		}
		return null; 
	}

	public void delete(Integer id) {
		measureRepository.deleteById(id);
	}
}

File: medicalcheck/app/src/main/java/ch/hearc/medicalcheck/services/MeasureService.kt
package ch.hearc.medicalcheck.services

import ch.hearc.medicalcheck.models.Measure
import ch.hearc.medicalcheck.models.User
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Response
import retrofit2.http.*

/*
* Project   : Medical Check Mobile
* Authors   : William Bikuta, Milán Cerviño, Ilyas Boillat, David Oktay
* Date      : 28.01.2022
* Class     : INF3dlm-a
* */

interface  MeasureService {
    @GET("measures")
    fun getAll(): Call<List<Measure>>

    @GET("measures/{id}")
    fun get(@Path("id") id : Int): Call<Measure>

    @GET("measures/user/{iduser}")
    fun getAllByIdUser(@Path("iduser") iduser: Int): Call<List<Measure>>

    @GET("measures/user/last/{iduser}")
    fun getLastByIdUser(@Path("iduser") iduser: Int): Call<Measure>

    @GET("measures/user/average/{iduser}/{date}")
    fun getMapAverage(@Path("iduser") iduser: Int?, @Path("date") date: String?): Call<Map<Int, Double>>

    @POST("measures")
    fun create(@Body measure: Measure): Call<Measure>

    @DELETE("measures/{id}")
    fun delete(@Path("id") id: Int): Call<ResponseBody>

    @PUT("measures/{id}")
    fun update(@Path("id") id : Int, @Body measure: Measure): Call<Measure>
}